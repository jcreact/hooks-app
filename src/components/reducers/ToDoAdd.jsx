import React, { useRef } from 'react';
import PropTypes from 'prop-types';

// Hooks
import { useForm } from '../../hooks/useForm';

/**
 * Formulario para agregar una nueva tarea a la lista.
 * @param {Object} props - Propiedades del componente.
 * @param {(todo: {id: number, project:string, task:string, done:boolean}) => void} props.handleAdd - Referencia a la función para agregar una tarea.
 */
export const ToDoAdd = ({ handleAdd }) => {
    const [{ task, project }, handleInputChange, reset] = useForm({
        task: '',
        project: 'Personal',
    });

    /**
     * @param {React.FormEvent<HTMLFormElement>} evt
     */
    const handleSubmint = (evt) => {
        evt.preventDefault();

        if (!task || !task.trim()) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            project: project.trim() || 'Personal',
            task,
            done: false,
        };

        handleAdd(newTodo);
        reset();
        document.querySelector(`input[name='task']`)?.select();
    };

    return (
        <form className="mb-3" onSubmit={handleSubmint}>
            <fieldset>
                <div className="form-floating">
                    <input
                        type="text"
                        className="form-control"
                        name="task"
                        placeholder="..."
                        value={task}
                        onChange={handleInputChange}
                        autoComplete="off"
                        autoFocus
                    />
                    <label>New Task</label>
                </div>

                <div className="input-group mb-2"></div>
                <div id="collapseInfo" className="">
                    <div className="input-group mb-2">
                        <span className="input-group-text">Project Name</span>
                        <input
                            type="text"
                            className="form-control"
                            name="project"
                            value={project}
                            onChange={handleInputChange}
                            autoComplete="off"
                        />
                        <button
                            type="submit"
                            className="btn btn-primary shadow-none"
                            disabled={!task}
                        >
                            <i className="bi bi-journal-plus me-2"></i>
                            Add
                        </button>
                    </div>

                    <div className="d-flex justify-content-end"></div>
                </div>
            </fieldset>
        </form>
    );
};

// Especificación de tipos de las propiedades.
ToDoAdd.propTypes = {
    handleAdd: PropTypes.func.isRequired,
};
