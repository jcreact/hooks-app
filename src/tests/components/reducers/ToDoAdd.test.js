import { shallow } from 'enzyme';

//Componente
import { ToDoAdd } from '../../../components/reducers/ToDoAdd.jsx';

describe(`Pruebas en el componente 'ToDoList.jsx'.`, () => {

    const handleAdd = jest.fn();
    const component = shallow(
        <ToDoAdd handleAdd={handleAdd} />
    );

    test(`Debe renderizar el componente ToDoListItem correctamente.`, () => {
        expect(component).toMatchSnapshot();
    });

    test(`No debe se ejecutarse la función 'handleAdd' si el formulario está vacío`, () => {

        const formSubmit = component.find('form').prop('onSubmit');

        formSubmit({ preventDefault() { } });
        expect(handleAdd).toHaveBeenCalledTimes(0);
    });

    test(`Debe ejecutar la función 'handleAdd' y restablecer el formulario.`, () => {

        const todo = {
            id: expect.any(Number),
            project: 'Demo',
            task: 'Tarea 1',
            done: false
        };

        component.find(`input[name='task']`).simulate('change', {
            target: {
                value: todo.task,
                name: 'task'
            }
        });
        component.find(`input[name='project']`).simulate('change', {
            target: {
                value: todo.project,
                name: 'project'
            }
        });

        const formSubmit = component.find('form').prop('onSubmit');
        formSubmit({ preventDefault() { } });

        expect(handleAdd).toBeCalled();
        expect(handleAdd).toHaveBeenCalledWith(todo);

        // Revisa que se haya restablecido el formulario.
        expect(component.find(`input[name='task']`).prop('value')).toBe('');
        expect(component.find(`input[name='project']`).prop('value')).toBe('Personal');
    });

});
