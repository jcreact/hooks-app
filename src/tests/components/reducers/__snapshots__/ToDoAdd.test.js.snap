// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente 'ToDoList.jsx'. Debe renderizar el componente ToDoListItem correctamente. 1`] = `
<form
  className="mb-3"
  onSubmit={[Function]}
>
  <fieldset>
    <div
      className="form-floating"
    >
      <input
        autoComplete="off"
        autoFocus={true}
        className="form-control"
        name="task"
        onChange={[Function]}
        placeholder="..."
        type="text"
        value=""
      />
      <label>
        New Task
      </label>
    </div>
    <div
      className="input-group mb-2"
    />
    <div
      className=""
      id="collapseInfo"
    >
      <div
        className="input-group mb-2"
      >
        <span
          className="input-group-text"
        >
          Project Name
        </span>
        <input
          autoComplete="off"
          className="form-control"
          name="project"
          onChange={[Function]}
          type="text"
          value="Personal"
        />
        <button
          className="btn btn-primary shadow-none"
          disabled={true}
          type="submit"
        >
          <i
            className="bi bi-journal-plus me-2"
          />
          Add
        </button>
      </div>
      <div
        className="d-flex justify-content-end"
      />
    </div>
  </fieldset>
</form>
`;
